### PR TITLE
James 1927: Override Cassandra Guice server with a LDAP user repository.

### PR DESCRIPTION
**Pull request by quynhn**

### CODE REVIEW

**Comment by quynhn** [comment]
test this please

File: server/container/guice/cassandra-ldap-guice/pom.xml [hunk 1]
```diff
@@ -0,0 +1,264 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (chibenwa) [inline server/container/guice/cassandra-ldap-guice/pom.xml]: Commit name should contain ticket number

File: server/container/guice/cassandra-ldap-guice/src/main/java/org/apache/james/CassandraLdapJamesServerMain.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+import org.apache.james.data.LdapUsersRepositoryModule;
+import org.apache.james.modules.server.JMXServerModule;
+
+import com.google.inject.Module;
+import com.google.inject.util.Modules;
+
+public class CassandraLdapJamesServerMain {
+
+    public static final Module cassandraServerModule = Modules.override(CassandraJamesServerMain.cassandraServerModule)
```
Comment (chibenwa) [inline server/container/guice/cassandra-ldap-guice/src/main/java/org/apache/james/CassandraLdapJamesServerMain.java]: cassandraLdapServerModule

File: server/container/guice/cassandra-ldap-guice/pom.xml [hunk 3]
```diff
@@ -193,12 +193,144 @@
...
+                    <scope>test</scope>
+                </dependency>
+                <dependency>
+                    <groupId>org.testcontainers</groupId>
+                    <artifactId>testcontainers</artifactId>
+                    <scope>test</scope>
+                </dependency>
+                <dependency>
+                    <groupId>${project.groupId}</groupId>
+                    <artifactId>james-server-cassandra-guice</artifactId>
```
Comment (aduprat) [inline server/container/guice/cassandra-ldap-guice/pom.xml]: Please group James dependencies first (in alphabetic order)
then external dependencies (in alphabetic order).
All of this without taking care of scope.

File: server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJmapTestRule.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            configuration.addProperty("[@userIdAttribute]", "uid");
+            configuration.addProperty("[@userObjectClass]", "inetOrgPerson");
+            configuration.addProperty("[@maxRetries]", "4");
+            configuration.addProperty("[@retryStartInterval]", "0");
+            configuration.addProperty("[@retryMaxInterval]", "8");
+            configuration.addProperty("[@retryIntervalScale]", "1000");
+            return configuration;
+        }
+    }
+
```
Comment (chibenwa) [inline server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJmapTestRule.java]: extra blank lines

File: server/container/guice/cassandra-ldap-guice/src/main/java/org/apache/james/data/LdapUsersRepositoryModule.java [hunk 5]
```diff
@@ -74,7 +73,7 @@ public void initModule() {
 
         @Override
         public List<Class<? extends Configurable>> forClasses() {
-            return ImmutableList.of(CassandraUsersRepository.class);
+            return ImmutableList.of(ReadOnlyUsersLDAPRepository.class);
```
Comment (aduprat) [inline server/container/guice/cassandra-ldap-guice/src/main/java/org/apache/james/data/LdapUsersRepositoryModule.java]: Should be in first commit.

File: server/container/guice/cassandra-ldap-guice/src/main/java/org/apache/james/CassandraLdapJamesServerMain.java [hunk 6]
```diff
@@ -27,13 +27,11 @@
...
 public class CassandraLdapJamesServerMain {
 
-    public static final Module cassandraServerModule = Modules.override(CassandraJamesServerMain.cassandraServerModule)
-            .with(new LdapUsersRepositoryModule());
-
-
     public static void main(String[] args) throws Exception {
         GuiceJamesServerImpl server = new GuiceJamesServerImpl()
-                    .combineWith(cassandraServerModule, new JMXServerModule());
+            .combineWith(CassandraJamesServerMain.cassandraServerModule, new JMXServerModule())
```
Comment (aduprat) [inline server/container/guice/cassandra-ldap-guice/src/main/java/org/apache/james/CassandraLdapJamesServerMain.java]: Should be in first commit.

File: server/container/guice/cassandra-ldap-guice/src/main/java/org/apache/james/data/LdapUsersRepositoryModule.java (hunk 5 above)
Comment (chibenwa) [inline server/container/guice/cassandra-ldap-guice/src/main/java/org/apache/james/data/LdapUsersRepositoryModule.java]: Squash this with the first commit

File: server/container/guice/cassandra-ldap-guice/src/main/java/org/apache/james/CassandraLdapJamesServerMain.java [hunk 7]
```diff
@@ -27,13 +27,11 @@
 
 public class CassandraLdapJamesServerMain {
 
-    public static final Module cassandraServerModule = Modules.override(CassandraJamesServerMain.cassandraServerModule)
-            .with(new LdapUsersRepositoryModule());
-
-
     public static void main(String[] args) throws Exception {
         GuiceJamesServerImpl server = new GuiceJamesServerImpl()
```
Comment (chibenwa) [inline server/container/guice/cassandra-ldap-guice/src/main/java/org/apache/james/CassandraLdapJamesServerMain.java]: Squash with previous commit

File: server/container/guice/cassandra-ldap-guice/pom.xml [hunk 8]
```diff
@@ -193,12 +193,144 @@
...
+                    <version>${assertj-3.version}</version>
+                    <scope>test</scope>
+                </dependency>
+                <dependency>
+                    <groupId>org.cassandraunit</groupId>
+                    <artifactId>cassandra-unit</artifactId>
+                    <version>${cassandra-unit.version}</version>
+                    <scope>test</scope>
+                </dependency>
+                <dependency>
```
Comment (chibenwa) [inline server/container/guice/cassandra-ldap-guice/pom.xml]: Why mockito is needed ?

File: server/container/guice/cassandra-ldap-guice/pom.xml [hunk 9]
```diff
@@ -193,12 +193,144 @@
...
+                    <groupId>${project.groupId}</groupId>
+                    <artifactId>james-server-guice-common</artifactId>
+                </dependency>
+                <dependency>
+                    <groupId>${project.groupId}</groupId>
+                    <artifactId>james-server-guice-common</artifactId>
+                    <type>test-jar</type>
+                    <scope>test</scope>
+                </dependency>
+                <dependency>
```
Comment (chibenwa) [inline server/container/guice/cassandra-ldap-guice/pom.xml]: Why this import is needed ?

File: server/container/guice/cassandra-ldap-guice/pom.xml [hunk 10]
```diff
@@ -193,12 +193,144 @@
...
+                    <groupId>org.apache.james</groupId>
+                    <artifactId>metrics-es-reporter</artifactId>
+                    <version>${project.version}</version>
+                </dependency>
+                <dependency>
+                    <groupId>${project.groupId}</groupId>
+                    <artifactId>james-server-guice-common</artifactId>
+                </dependency>
+                <dependency>
```
Comment (chibenwa) [inline server/container/guice/cassandra-ldap-guice/pom.xml]: Why this import is needed ? Can we remove it ?

File: server/container/guice/cassandra-ldap-guice/pom.xml [hunk 11]
```diff
@@ -193,12 +193,144 @@
...
+                    <groupId>${project.groupId}</groupId>
+                    <artifactId>james-server-data-cassandra</artifactId>
+                </dependency>
+                <dependency>
+                    <groupId>org.apache.james</groupId>
+                    <artifactId>james-server-data-jmap-cassandra</artifactId>
+                </dependency>
+                <dependency>
```
Comment (chibenwa) [inline server/container/guice/cassandra-ldap-guice/pom.xml]: Why this import is needed ? Can we remove it ?

File: server/container/guice/cassandra-ldap-guice/pom.xml [hunk 12]
```diff
@@ -193,12 +193,144 @@
...
+                    <groupId>${project.groupId}</groupId>
+                    <artifactId>apache-james-mailbox-tika</artifactId>
+                </dependency>
+                <dependency>
+                    <groupId>${project.groupId}</groupId>
+                    <artifactId>james-server-data-cassandra</artifactId>
+                </dependency>
+                <dependency>
```
Comment (chibenwa) [inline server/container/guice/cassandra-ldap-guice/pom.xml]: Why this import is needed ? Can we remove it ?

File: server/container/guice/cassandra-ldap-guice/pom.xml [hunk 13]
```diff
@@ -193,12 +193,144 @@
...
+                    <groupId>org.apache.james</groupId>
+                    <artifactId>apache-mailet-icalendar</artifactId>
+                    <version>${project.version}</version>
+                </dependency>
+                <dependency>
+                    <groupId>${project.groupId}</groupId>
+                    <artifactId>apache-james-mailbox-tika</artifactId>
+                </dependency>
+                <dependency>
```
Comment (chibenwa) [inline server/container/guice/cassandra-ldap-guice/pom.xml]: Why this import is needed ? Can we remove it ?

File: server/container/guice/cassandra-ldap-guice/pom.xml [hunk 14]
```diff
@@ -193,12 +193,144 @@
...
+                    <artifactId>apache-james-mailbox-elasticsearch</artifactId>
+                    <type>test-jar</type>
+                    <scope>test</scope>
+                </dependency>
+                <dependency>
+                    <groupId>org.apache.james</groupId>
+                    <artifactId>apache-mailet-icalendar</artifactId>
+                    <version>${project.version}</version>
+                </dependency>
+                <dependency>
```
Comment (chibenwa) [inline server/container/guice/cassandra-ldap-guice/pom.xml]: Why this import is needed ? Can we remove it ?

File: server/container/guice/cassandra-ldap-guice/pom.xml [hunk 15]
```diff
@@ -193,12 +193,144 @@
...
+                    <groupId>${project.groupId}</groupId>
+                    <artifactId>apache-james-mailbox-elasticsearch</artifactId>
+                </dependency>
+                <dependency>
+                    <groupId>${project.groupId}</groupId>
+                    <artifactId>apache-james-mailbox-elasticsearch</artifactId>
+                    <type>test-jar</type>
+                    <scope>test</scope>
+                </dependency>
+                <dependency>
```
Comment (chibenwa) [inline server/container/guice/cassandra-ldap-guice/pom.xml]: Why this import is needed ? Can we remove it ?

File: server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJamesServerTest.java [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    @Override
+    protected void clean() {
+        if (ldapContainer != null) {
+            ldapContainer.stop();
+        }
+    }
+
+    @Test
+    public void userFromLdapShouldLoginViaImapProtocol() throws Exception {
+        IMAPClient imapClient = new IMAPClient();
```
Comment (aduprat) [inline server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJamesServerTest.java]: You should disconnect the client at the end of the test.

File: server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJmapTestRule.java [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+    public CassandraLdapJmapTestRule(GuiceModuleTestRule... guiceModuleTestRule) {
+        this.guiceModuleTestRule =
+            AggregateGuiceModuleTestRule
+                .of(guiceModuleTestRule)
+                .aggregate(new TempFilesystemTestRule());
+    }
+
+    public JmapJamesServer jmapServer(String ldapIp, Module... additionals) {
+        return new JmapJamesServer()
+            .combineWith(CassandraJamesServerMain.cassandraServerModule,
```
Comment (aduprat) [inline server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJmapTestRule.java]: You should use CassandraLdapJamesServerMain

**Comment by aduprat** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

File: server/container/guice/cassandra-ldap-guice/pom.xml (hunk 1 above)
Comment (quynhn) [inline server/container/guice/cassandra-ldap-guice/pom.xml]: Yes, I changed the history

File: server/container/guice/cassandra-ldap-guice/pom.xml (hunk 15 above)
Comment (quynhn) [inline server/container/guice/cassandra-ldap-guice/pom.xml]: I removed the redundant dependencies and reorder it

File: server/container/guice/cassandra-ldap-guice/src/main/java/org/apache/james/CassandraLdapJamesServerMain.java (hunk 7 above)
Comment (quynhn) [inline server/container/guice/cassandra-ldap-guice/src/main/java/org/apache/james/CassandraLdapJamesServerMain.java]: I changed the history than squash later

File: server/container/guice/cassandra-ldap-guice/src/main/java/org/apache/james/CassandraLdapJamesServerMain.java (hunk 2 above)
Comment (quynhn) [inline server/container/guice/cassandra-ldap-guice/src/main/java/org/apache/james/CassandraLdapJamesServerMain.java]: It's removed later

File: server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJmapTestRule.java (hunk 4 above)
Comment (quynhn) [inline server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJmapTestRule.java]: Done

File: server/container/guice/cassandra-ldap-guice/pom.xml (hunk 3 above)
Comment (quynhn) [inline server/container/guice/cassandra-ldap-guice/pom.xml]: Yes, I did it in the last commit

File: server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJamesServerTest.java (hunk 16 above)
Comment (quynhn) [inline server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJamesServerTest.java]: OK, I've done

File: server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJmapTestRule.java (hunk 17 above)
Comment (quynhn) [inline server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJmapTestRule.java]: I reuse the cassandraServerModule such it still cassandra then override UserRepo by Ldap.
Is that ok or I have to duplicate that variable from CassandraJamesServerMain to CassandraLdapJamesServerMain?

File: server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJamesServerTest.java [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+        imapClient.connect(JAMES_SERVER_HOST, IMAP_PORT);
+
+        assertThat(imapClient.login(JAMES_USER, PASSWORD)).isTrue();
+        imapClient.disconnect();
```
Comment (chibenwa) [inline server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJamesServerTest.java]: No. If the test fails you will not be disconnected. This should be either in tearDown or in a try with resources statement.

File: server/container/guice/cassandra-ldap-guice/src/main/java/org/apache/james/CassandraLdapJamesServerMain.java [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+import org.apache.james.modules.server.JMXServerModule;
+
+import com.google.inject.Module;
+import com.google.inject.util.Modules;
+
+public class CassandraLdapJamesServerMain {
+
+    public static void main(String[] args) throws Exception {
+        GuiceJamesServerImpl server = new GuiceJamesServerImpl()
+            .combineWith(CassandraJamesServerMain.cassandraServerModule, new JMXServerModule())
```
Comment (chibenwa) [inline server/container/guice/cassandra-ldap-guice/src/main/java/org/apache/james/CassandraLdapJamesServerMain.java]: To handle @aduprat comment, could you introduce cassandraLdapModule static field defined as : 

```
Modules.override(CassandraJamesServerMain.cassandraServerModule).with(new LdapUsersRepositoryModule());
```

Then you can simply use : 

```
GuiceJamesServerImpl server = new GuiceJamesServerImpl()
            .combineWith(cassandraLdapModule, new JMXServerModule());
```

Directly.

File: server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJmapTestRule.java (hunk 17 above)
Comment (chibenwa) [inline server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJmapTestRule.java]: With the proposition made above, taking care of @aduprat comment should be OK

File: server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJamesServerTest.java (hunk 18 above)
Comment (quynhn) [inline server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJamesServerTest.java]: I move it to tearDown

**Comment by mbaechler** [comment]
could we have clean branch to review now ?

**Comment by quynhn** [comment]
```
could we have clean branch to review now ?
```
Yes, I've just squashed to make it better for review

File: server/container/guice/cassandra-ldap-guice/pom.xml [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+                    <groupId>${project.groupId}</groupId>
+                    <artifactId>james-server-util-java8</artifactId>
+                    <scope>test</scope>
+                    <type>test-jar</type>
+                </dependency>
+                <dependency>
+                    <groupId>com.jayway.awaitility</groupId>
+                    <artifactId>awaitility</artifactId>
```
Comment (mbaechler) [inline server/container/guice/cassandra-ldap-guice/pom.xml]: should it be in test scope ?

**Pull request merged**